src/state/service.rs

```diff
@@ -39,6 +41,10 @@ pub struct Service {
     /// Maps endpoint UIDs to service [Endpoint]s.
     #[serde(default)]
     pub endpoints: HashMap<String, Endpoint>,
+    /// Cache the set of gateways in front of this Service
+    /// based on its endoints.
+    #[serde(default)]
+    pub waypoints: HashMap<Destination, usize>,
```

Comment:
An alternative here is to denormalize the `waypoint` field of `Workload` into the `Endpoint` struct. Lookups aren't quite as fast though.